<a href="https://colab.research.google.com/github/meghnabajoria/Diabetes-prediction-using-support-vector-machine/blob/main/hyperparameter_tuning_Diabetes_prediction_using_support_vector_machine.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
# importing libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import svm
from sklearn.metrics import accuracy_score

In [6]:
# read csv
data = pd.read_csv('/content/diabetes.csv')
# NOTE : you can use pd.read_csv? if you dont know what it does or what are its parameters.
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [7]:
isinstance(data, pd.DataFrame)

True

In [8]:
# get to know the data
print("Dimension of the data", data.shape)
data.info()

Dimension of the data (768, 9)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


There are 768 rows which means this data has been collected from 768 people. Since number of rows in the dataset is same as non-null values in each column, we can conclude that there are no null values. However, let's verify 

In [9]:
# statistical measures of the data
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [10]:
data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

Hence, we can conclude there are no missing values.

In [11]:
# let's see how many diabetic and non-diabetic patients are there
data['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

Label 0 represents people without diabets and label 1 represents people without diabetes. Even though the data is imbalanced, we will ignore it since this is a small dataset.

In [12]:
data.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

Here, we notice that the glucose, BMI and Insulin levels and age of people with diabetes is higher.

In [13]:
# Seperating the data and the outcome
X = data.drop(columns = 'Outcome', axis = 1)
Y = data['Outcome']

In [14]:
print(X)
print(Y)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


Now we have seperated the features and the labels.

In [15]:
# Data standardization
scaler = StandardScaler()

In [16]:
scaler.fit(X)

StandardScaler()

In [17]:
standardized_data = scaler.transform(X)

In [18]:
print(standardized_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


We can now see that all the values are in similar range i.e from -1 to 1.

In [19]:
X = standardized_data
Y = data['Outcome']
print(X)
print(Y)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]
0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


In [20]:
# now lets split our data into training and test data
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify = Y, random_state = 1)

In [21]:
# training the model
classifier = svm.SVC(kernel = 'linear')

#training the support vector machine classifer
classifier.fit(X_train, Y_train)

SVC(kernel='linear')

Now, lets evaluate our model

In [22]:
X_train_prediction = classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)
print('Accuracy score of the training data', training_data_accuracy)

Accuracy score of the training data 0.7833876221498371


In [23]:
# let's evaluate the model on test data
X_test_prediction = classifier.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)
print('Accuracy score of the test data', test_data_accuracy)

Accuracy score of the test data 0.7792207792207793


In [38]:
accuracy = metrics.accuracy_score(X_test_prediction, Y_test)
precision = metrics.precision_score(X_test_prediction, Y_test)
recall = metrics.recall_score(X_test_prediction, Y_test)
f1_score = metrics.f1_score(X_test_prediction, Y_test)
print(accuracy)
print(precision)
print(recall)
print(f1_score)

0.7792207792207793
0.5370370370370371
0.7631578947368421
0.6304347826086957


The model seems to perform equally well on the test data too which shows that there is no overfitting.

In [24]:
# making a predictive system
input_data = (4,110,92,0,0,37.6,0.191,30)

# change the input data to np array
input_data_as_np = np.asarray(input_data)

# reshape the array
input_data_reshaped = input_data_as_np.reshape(1,-1)

# standardized the data
std_data = scaler.transform(input_data_reshaped)

print(std_data)

[[ 0.04601433 -0.34096773  1.18359575 -1.28821221 -0.69289057  0.71168975
  -0.84827977 -0.27575966]]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [25]:
# Now let's make a prediction
prediction = classifier.predict(std_data)
print(prediction)

if(prediction == 0):
    print("The person is non-diabetic")
else:
    print("The person is diabetic")

[0]
The person is non-diabetic


HYPERPARAMETER TUNING

In [30]:
from sklearn import svm
from sklearn.model_selection import RandomizedSearchCV

C = [float(x) for x in np.linspace(start=0.1, stop=100, num=10)]
gamma = [float(x) for x in np.linspace(start=0.1, stop=1, num=10)]
kernels = ['rbf', 'linear', 'poly', 'sigmoid', 'precomputed']
valid_kernels = []

for kernel in kernels:
    try:
        classifier = svm.SVC(kernel=kernel)
        classifier.fit(X_train, Y_train)
        valid_kernels.append(kernel)
    except ValueError:
        pass

degree = list(range(1, 6))
coef0 = [float(x) for x in np.linspace(0, 1, 10)]
shrinking = [True, False]

random_grid = {'C': C,
               'gamma': gamma,
               'kernel': valid_kernels,
               'degree': degree,
               'coef0': coef0,
               'shrinking': shrinking}
print(random_grid)


{'C': [0.1, 11.200000000000001, 22.300000000000004, 33.400000000000006, 44.50000000000001, 55.60000000000001, 66.7, 77.80000000000001, 88.9, 100.0], 'gamma': [0.1, 0.2, 0.30000000000000004, 0.4, 0.5, 0.6, 0.7000000000000001, 0.8, 0.9, 1.0], 'kernel': ['rbf', 'linear', 'poly', 'sigmoid'], 'degree': [1, 2, 3, 4, 5], 'coef0': [0.0, 0.1111111111111111, 0.2222222222222222, 0.3333333333333333, 0.4444444444444444, 0.5555555555555556, 0.6666666666666666, 0.7777777777777777, 0.8888888888888888, 1.0], 'shrinking': [True, False]}


In [31]:
svm = svm.SVC()
svm_randomcv = RandomizedSearchCV(svm, param_distributions=random_grid, n_iter=10, cv=5, verbose=2, random_state=42)
svm_randomcv.fit(X_train, Y_train)

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[CV] END C=33.400000000000006, coef0=1.0, degree=3, gamma=0.5, kernel=linear, shrinking=False; total time=   0.5s
[CV] END C=33.400000000000006, coef0=1.0, degree=3, gamma=0.5, kernel=linear, shrinking=False; total time=   0.2s
[CV] END C=33.400000000000006, coef0=1.0, degree=3, gamma=0.5, kernel=linear, shrinking=False; total time=   0.6s
[CV] END C=33.400000000000006, coef0=1.0, degree=3, gamma=0.5, kernel=linear, shrinking=False; total time=   0.2s
[CV] END C=33.400000000000006, coef0=1.0, degree=3, gamma=0.5, kernel=linear, shrinking=False; total time=   0.5s
[CV] END C=0.1, coef0=0.2222222222222222, degree=1, gamma=0.8, kernel=poly, shrinking=True; total time=   0.0s
[CV] END C=0.1, coef0=0.2222222222222222, degree=1, gamma=0.8, kernel=poly, shrinking=True; total time=   0.0s
[CV] END C=0.1, coef0=0.2222222222222222, degree=1, gamma=0.8, kernel=poly, shrinking=True; total time=   0.0s
[CV] END C=0.1, coef0=0.222222222222

RandomizedSearchCV(cv=5, estimator=SVC(),
                   param_distributions={'C': [0.1, 11.200000000000001,
                                              22.300000000000004,
                                              33.400000000000006,
                                              44.50000000000001,
                                              55.60000000000001, 66.7,
                                              77.80000000000001, 88.9, 100.0],
                                        'coef0': [0.0, 0.1111111111111111,
                                                  0.2222222222222222,
                                                  0.3333333333333333,
                                                  0.4444444444444444,
                                                  0.5555555555555556,
                                                  0.6666666666666666,
                                                  0.7777777777777777,
                                                  0.8888888888888888, 1.0],
                                        'degree': [1, 2, 3, 4, 5],
                                        'gamma': [0.1, 0.2, 0.30000000000000004,
                                                  0.4, 0.5, 0.6,
                                                  0.7000000000000001, 0.8, 0.9,
                                                  1.0],
                                        'kernel': ['rbf', 'linear', 'poly',
                                                   'sigmoid'],
                                        'shrinking': [True, False]},
                   random_state=42, verbose=2)

In [32]:
print(svm_randomcv.best_params_)

{'shrinking': True, 'kernel': 'poly', 'gamma': 0.8, 'degree': 1, 'coef0': 0.2222222222222222, 'C': 0.1}


In [34]:
best_svm = svm_randomcv.best_estimator_
best_svm

SVC(C=0.1, coef0=0.2222222222222222, degree=1, gamma=0.8, kernel='poly')

In [36]:
from sklearn import metrics
y_pred = best_svm.predict(X_test)


accuracy = metrics.accuracy_score(Y_test, y_pred)
precision = metrics.precision_score(Y_test, y_pred)
recall = metrics.recall_score(Y_test, y_pred)
f1_score = metrics.f1_score(Y_test, y_pred)
print(accuracy)
print(precision)
print(recall)
print(f1_score)

0.7727272727272727
0.7567567567567568
0.5185185185185185
0.6153846153846154


We saw no significant improvemet in accuracy.